operator/pkg/ciliumendpointbatch/endpointbatch.go

```diff
@@ -0,0 +1,321 @@
+
+	cebC := &CiliumEndpointBatchController{
+		clientV2:                            client.CiliumV2(),
+		clientV2a1:                          client.CiliumV2alpha1(),
+		aggregator:                          aggregator,
+		reconciler:                          newReconciler(client.CiliumV2alpha1(), manager),
+		Manager:                             manager,
+		cebDeleteSyncTime:                   0,
+		ciliumEndpointBatchUpsertSyncPeriod: cebSyncPeriod,
+		ciliumEndpointBatchDeleteSyncPeriod: cebDeleteSyncPeriod,
+	}
+
+	// List all existing CEBs from API server and cache it locally.
+	// This sync should happen before starting CEP watcher, because CEP watcher
+	// notifies the existing CEPs as newly added CEPs. If we don't have local sync
+	// cebManager would assume those are new CEPs and may create new CEBs for those CEPs.
+	// This situation ends up having duplicate CEPs in different CEBs. Hence, we need
+	// to sync existing CEBs before starting a CEP watcher.
+	// On Error, return nil to caller.
+	if err := getCiliumEndpointBatchesFromApiServer(cebC.clientV2a1, cebC.Manager); err != nil {
+		log.WithError(err).Fatal("Failed to get CEBs from K8sAPIServer")
+		return nil
+	}
+
+	return cebC
+}
+
+// reconcileCeb gets the list of CEB's to be created, updated and deleted
+// from cebAggregator. It makes an attempt to reconcile above listed CEB's
+// with k8s-apiserver.
+// If reconciler have any issues or errors while syncing with k8s-apiserver,
+// corresponding CEBs are queued back to aggregator to sync with API server
+// in the next attempt.
+func (c *CiliumEndpointBatchController) reconcileCeb() (bool, error) {
+
+	var errCebCreate, errCebUpdate, errCebDelete []string
+	// If the current latency cycle exceeds the cebDeleteSyncPeriod,
+	// resync all CEB's, Includes Create, Update and Delete.
+	if c.cebDeleteSyncTime >= c.ciliumEndpointBatchDeleteSyncPeriod {
+		c.cebDeleteSyncTime = 0
+		errCebCreate, errCebUpdate, errCebDelete =
+			c.reconciler.reconcileWithServer(c.aggregator.getCreateUpdateAndDeleteCebNames())
+	} else {
+		// During normal reconcile, only CebCreate and CebUpdate are reconciled.
+		// pass an empty cebDelete to reconciler.
+		var cebDelete []string
+		cebCreate, cebUpdate := c.aggregator.getCreateAndUpdateCebNames()
+		c.cebDeleteSyncTime += c.ciliumEndpointBatchUpsertSyncPeriod
+		// Reconcile only Create and Update CEB's with k8s-apiserver.
+		errCebCreate, errCebUpdate, errCebDelete =
+			c.reconciler.reconcileWithServer(cebCreate, cebUpdate, cebDelete)
+	}
+
+	// If local CEB's are not reconciled with k8s-apiserver,
+	// queue them in aggregator to resync during next reconcile process.
+	for _, cebName := range errCebCreate {
+		c.aggregator.updateAggregator(cebName, CebCreate)
+	}
+
+	for _, cebName := range errCebUpdate {
+		c.aggregator.updateAggregator(cebName, CebUpdate)
+	}
+
+	for _, cebName := range errCebDelete {
+		c.aggregator.updateAggregator(cebName, CebDelete)
+	}
```

Comment:
@aanm Current code, doesn't handle the scenario you have mentioned. Let me state the problem again to make sure we are on the same page. 

If k8s-apiserver re-creates CEP with exact same values as the past one [which died shortly]. The sequence of events are like below.,
1.  CEP-A-CREATE
2. CEP-A-DELETE
3. CEP-XYZ-UPDATE
4. CEP-A-CREATE

How does batching code runs in operator ensures this order to cilium-agent ? Unfortunately this PR doesn't guarantee that CEP-A-DELETE in step 2 goes before CEP-A-CREATE in step 4. It can't guarantee that unless they are in the CEB.  Let me think about how we can solve this issue along with mentioned [here](https://github.com/cilium/cilium/issues/16565) . Correct me if didn't state the issue you are mentioning.

Reply:
As long as if `CEP`s are recreated with same `cepName` and `cepNamespace`, underlying Issue is solved with latest code. However following issue isn't addressed in this batching code.
1) If re-generated CEPs are created with exact same value as one of the CEP died earlier. That means for the re-generated CEPs Cilium-agent, might receive 'CEP-DELETE' request after 'CEP-INSERT' . Cilium agent should be smart enough to figure this one.